core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,183 @@
+
+  /**
+   * Top level search workflow
+   *
+   * @throws IOException
+   */
+  @Override
+  public void process() throws IOException {
+    List<String> matchedLines = new ArrayList<String>();
+    for (File file : listFiles(getRootPath())) {
+      for (String line : readLines(file)) {
+        if (containsPattern(line)) {
+          matchedLines.add(line);
+        }
+      }
+    }
+    writeToFile(matchedLines);
+  }
+
+  /**
+   * Traverse a given directory and return all files
+   *
+   * @param rootDir input Directory
+   * @return files under the root directory
+   */
+  @Override
+  public List<File> listFiles(String rootDir) throws IOException {
+    List<File> files = new ArrayList<File>();
+    Queue<File> fileQueue = new LinkedList<>();
+    fileQueue.add(new File(rootDir));
+    while (!fileQueue.isEmpty()) {
+      File currentFile = fileQueue.poll();
+      if (currentFile.isDirectory()) {
+        for (File file : currentFile.listFiles()) {
+          fileQueue.add(file);
+        }
+      } else {
+        files.add(currentFile);
+      }
+    }
+    return files;
+  }
+
+  /**
+   * Read a file and return all the lines
+   *
+   * @param inputFile file to be read
+   * @return lines
+   * @throws IllegalArgumentException if a given input file is not a file
+   */
+  @Override
+  public List<String> readLines(File inputFile) throws IOException {
+    if (!inputFile.isFile()) {
+      throw new IllegalArgumentException("ERROR: inputFile is not a file.");
+    }
+    List<String> lines = new ArrayList<>();
+    try {
+      BufferedReader bufferedReader = new BufferedReader(new FileReader(inputFile));
+      String line;
+      while ((line = bufferedReader.readLine()) != null) {
+        lines.add(line);
+      }
+    } catch (IOException ex) {
+      logger.error("ERROR: Failed to create BufferedReader", ex);
+    }
+    return lines;
+  }
+
+  /**
+   * check if the line contains the regex pattern (passed by the user)
+   *
+   * @param line input string
+   * @return true if there is a match
+   */
+  @Override
+  public boolean containsPattern(String line) {
+    Pattern pattern = Pattern.compile(
+        getRegex());
+    Matcher matcher = pattern.matcher(line);
+    boolean match = matcher.matches();
+    return match;
+  }
+
+  /**
+   * Write lines to a file
+   *
+   * @param lines matched lines
+   * @throws IOException if write failed
+   */
+  @Override
+  public void writeToFile(List<String> lines) throws IOException {
+    File outputFile = new File(getOutFile());
+    FileOutputStream outStream = new FileOutputStream(outputFile);
+    try {
+      BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(outStream));
+      for (String line : lines) {
+        bufferedWriter.write(line);
+        bufferedWriter.newLine();
+      }
+      bufferedWriter.flush();
+      bufferedWriter.close();
+    } catch (IOException ex) {
+      logger.error("ERROR: The write to outFile failed", ex);
+    }
```

Comment:
this method shall throw IOException if it's unable to write, so you don't need to catch the IOException

Sample implementation

```
  @Override
  public void writeToFile(List<String> lines) throws IOException {
    FileOutputStream fos = new FileOutputStream(this.getOutFile());
    OutputStreamWriter osw = new OutputStreamWriter(fos);
    BufferedWriter bw = new BufferedWriter(osw);

    for (String line : lines) {
      bw.write(line);
      bw.newLine();
    }
    //empty outFile is not close....
    bw.close();
  }
```

Reply:
Try catch has been removed